File: platform/core/event/src/main/java/com/android/event/ZdEvent.java
//package com.android.event;
//
//import android.arch.lifecycle.ExternalLiveData;
//import android.content.Context;
//import android.content.Intent;
//import android.content.IntentFilter;
//import android.os.Build;
//import android.os.Handler;
//import android.os.Looper;
//
//import androidx.annotation.MainThread;
//import androidx.annotation.NonNull;
//import androidx.annotation.Nullable;
//import androidx.lifecycle.Lifecycle;
//import androidx.lifecycle.LifecycleOwner;
//import androidx.lifecycle.Observer;
//
//import com.android.event.ipc.IpcConst;
//import com.android.event.ipc.encode.IEncode;
//import com.android.event.ipc.encode.ValueEncode;
//import com.android.event.ipc.receiver.LebIpcReceiver;
//import com.android.utils.ThreadUtil;
//import com.google.gson.Gson;
//
//import java.util.HashMap;
//import java.util.Map;
//
///**
// * created by jiangshide on 2019-06-18.
// * email:18311271399@163.com
// */
//public final class ZdEvent {
//    private final Map<String, LiveEvent<?>> bus;
//
//    private ZdEvent() {
//        bus = new HashMap<>();
//    }
//
//    private static class SingletonHolder {
//        private static final ZdEvent DEFAULT_BUS = new ZdEvent();
//    }
//
//    public static ZdEvent get() {
//        return SingletonHolder.DEFAULT_BUS;
//    }
//
//    private Context appContext;
//    private boolean lifecycleObserverAlwaysActive = true;
//    private boolean autoClear = false;
//    private IEncode encoder = new ValueEncode();
//    private Config config = new Config();
//    private LebIpcReceiver receiver = new LebIpcReceiver();
//
//    public synchronized <T> Observable<T> with(String key, Class<T> type) {
//        return this.with(key, type, false);
//    }
//
//    public synchronized <T> Observable<T> with(String key, Class<T> type, boolean isOut) {
//        if (!bus.containsKey(key)) {
//            bus.put(key, new LiveEvent<>(key, isOut ? type : null));
//        }
//        return (Observable<T>) bus.get(key);
//    }
//
//    public Observable<Object> with(String key) {
//        return with(key, Object.class);
//    }
//
//
//    /**
//     * use the inner class Config to set params
//     * first of all, call config to get the Config instance
//     * then, call the method of Config to config ZdEvent
//     * call this method in Application.onCreate
//     */
//
//    public Config config() {
//        return config;
//    }
//
//    public class Config {
//
//        /**
//         * lifecycleObserverAlwaysActive
//         * set if then observer can always receive message
//         * true: observer can always receive message
//         * false: observer can only receive message when resumed
//         *
//         * @param active
//         * @return
//         */
//        public Config lifecycleObserverAlwaysActive(boolean active) {
//            lifecycleObserverAlwaysActive = active;
//            return this;
//        }
//
//        /**
//         * @param clear
//         * @return true: clear livedata when no observer observe it
//         * false: not clear livedata unless app was killed
//         */
//        public Config autoClear(boolean clear) {
//            autoClear = clear;
//            return this;
//        }
//
//        /**
//         * config broadcast
//         * only if you called this method, you can use broadcastValue() to send broadcast message
//         *
//         * @param context
//         * @return
//         */
//        public Config supportBroadcast(Context context) {
//            if (context != null) {
//                appContext = context.getApplicationContext();
//            }
//            if (appContext != null) {
//                IntentFilter intentFilter = new IntentFilter();
//                intentFilter.addAction(IpcConst.ACTION);
//                appContext.registerReceiver(receiver, intentFilter);
//            }
//            return this;
//        }
//    }
//
//    public interface Observable<T> {
//
//        /**
//         * 发送一个消息，支持前台线程、后台线程发送
//         *
//         * @param value
//         */
//        void post(T value);
//
//        /**
//         * 发送一个消息，支持前台线程、后台线程发送
//         * 需要跨进程、跨APP发送消息的时候调用该方法
//         *
//         * @param value
//         */
//        void broadcast(T value);
//
//        /**
//         * 延迟发送一个消息，支持前台线程、后台线程发送
//         *
//         * @param value
//         * @param delay 延迟毫秒数
//         */
//        void postDelay(T value, long delay);
//
//        /**
//         * 发送一个消息，支持前台线程、后台线程发送
//         * 需要跨进程、跨APP发送消息的时候调用该方法
//         *
//         * @param value
//         */
//        void broadcast(T value, boolean foreground);
//
//        /**
//         * 注册一个Observer，生命周期感知，自动取消订阅
//         *
//         * @param owner
//         * @param observer
//         */
//        void observe(@NonNull LifecycleOwner owner, @NonNull Observer<T> observer);
//
//        /**
//         * 注册一个Observer，生命周期感知，自动取消订阅
//         * 如果之前有消息发送，可以在注册时收到消息（消息同步）
//         *
//         * @param owner
//         * @param observer
//         */
//        void observeSticky(@NonNull LifecycleOwner owner, @NonNull Observer<T> observer);
//
//        /**
//         * 注册一个Observer
//         *
//         * @param observer
//         */
//        void observeForever(@NonNull Observer<T> observer);
//
//        /**
//         * 注册一个Observer
//         * 如果之前有消息发送，可以在注册时收到消息（消息同步）
//         *
//         * @param observer
//         */
//        void observeStickyForever(@NonNull Observer<T> observer);
//
//        /**
//         * 通过observeForever或observeStickyForever注册的，需要调用该方法取消订阅
//         *
//         * @param observer
//         */
//        void removeObserver(@NonNull Observer<T> observer);
//    }
//
//    private class LiveEvent<T> implements Observable<T> {
//
//        @NonNull
//        private final String key;
//        private Class<T> type;
//        private final LifecycleLiveData<T> liveData;
//        private final Map<Observer, ObserverWrapper<T>> observerMap = new HashMap<>();
//        private final Handler mainHandler = new Handler(Looper.getMainLooper());
//
//
//        LiveEvent(@NonNull String key, Class<T> type) {
//            this.key = key;
//            this.type = type;
//            this.liveData = new LifecycleLiveData<>();
//        }
//
//        @Override
//        public void post(T value) {
//            if (ThreadUtil.isMainThread()) {
//                postInternal(value);
//            } else {
//                mainHandler.post(new PostValueTask(value));
//            }
//        }
//
//        @Override
//        public void broadcast(T value) {
//            broadcast(value, false);
//        }
//
//        @Override
//        public void postDelay(T value, long delay) {
//            mainHandler.postDelayed(new PostValueTask(value), delay);
//        }
//
//        @Override
//        public void broadcast(final T value, final boolean foreground) {
//            if (appContext != null) {
//                if (ThreadUtil.isMainThread()) {
//                    broadcastInternal(value, foreground);
//                } else {
//                    mainHandler.post(new Runnable() {
//                        @Override
//                        public void run() {
//                            broadcastInternal(value, foreground);
//                        }
//                    });
//                }
//            } else {
//                post(value);
//            }
//        }
//
//        @Override
//        public void observe(@NonNull final LifecycleOwner owner, @NonNull final Observer<T> observer) {
//            if (ThreadUtil.isMainThread()) {
//                observeInternal(owner, observer);
//            } else {
//                mainHandler.post(new Runnable() {
//                    @Override
//                    public void run() {
//                        observeInternal(owner, observer);
//                    }
//                });
//            }
//        }
//
//        @Override
//        public void observeSticky(@NonNull final LifecycleOwner owner, @NonNull final Observer<T> observer) {
//            if (ThreadUtil.isMainThread()) {
//                observeStickyInternal(owner, observer);
//            } else {
//                mainHandler.post(new Runnable() {
//                    @Override
//                    public void run() {
//                        observeStickyInternal(owner, observer);
//                    }
//                });
//            }
//        }
//
//        @Override
//        public void observeForever(@NonNull final Observer<T> observer) {
//            if (ThreadUtil.isMainThread()) {
//                observeForeverInternal(observer);
//            } else {
//                mainHandler.post(new Runnable() {
//                    @Override
//                    public void run() {
//                        observeForeverInternal(observer);
//                    }
//                });
//            }
//        }
//
//        @Override
//        public void observeStickyForever(@NonNull final Observer<T> observer) {
//            if (ThreadUtil.isMainThread()) {
//                observeStickyForeverInternal(observer);
//            } else {
//                mainHandler.post(new Runnable() {
//                    @Override
//                    public void run() {
//                        observeStickyForeverInternal(observer);
//                    }
//                });
//            }
//        }
//
//        @Override
//        public void removeObserver(@NonNull final Observer<T> observer) {
//            if (ThreadUtil.isMainThread()) {
//                removeObserverInternal(observer);
//            } else {
//                mainHandler.post(new Runnable() {
//                    @Override
//                    public void run() {
//                        removeObserverInternal(observer);
//                    }
//                });
//            }
//        }
//
//        @MainThread
//        private void postInternal(T value) {
//            liveData.setValue(value);
//        }
//
//        @MainThread
//        private void broadcastInternal(T value, boolean foreground) {
//            Intent intent = new Intent(IpcConst.ACTION);
//            if (foreground && Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
//                intent.addFlags(Intent.FLAG_RECEIVER_FOREGROUND);
//            }
//            intent.putExtra(IpcConst.KEY, key);
//            try {
//                encoder.encode(intent, value);
//                appContext.sendBroadcast(intent);
//            } catch (Exception e) {
//                e.printStackTrace();
//            }
//        }
//
//        @MainThread
//        private void observeInternal(@NonNull LifecycleOwner owner, @NonNull Observer<T> observer) {
//            ObserverWrapper<T> observerWrapper = new ObserverWrapper<>(observer, type);
//            observerWrapper.preventNextEvent = liveData.getVersion() > ExternalLiveData.START_VERSION;
//            liveData.observe(owner, observerWrapper);
//        }
//
//        @MainThread
//        private void observeStickyInternal(@NonNull LifecycleOwner owner, @NonNull Observer<T> observer) {
//            ObserverWrapper<T> observerWrapper = new ObserverWrapper<>(observer, type);
//            liveData.observe(owner, observerWrapper);
//        }
//
//        @MainThread
//        private void observeForeverInternal(@NonNull Observer<T> observer) {
//            ObserverWrapper<T> observerWrapper = new ObserverWrapper<>(observer, type);
//            observerWrapper.preventNextEvent = liveData.getVersion() > ExternalLiveData.START_VERSION;
//            observerMap.put(observer, observerWrapper);
//            liveData.observeForever(observerWrapper);
//        }
//
//        @MainThread
//        private void observeStickyForeverInternal(@NonNull Observer<T> observer) {
//            ObserverWrapper<T> observerWrapper = new ObserverWrapper<>(observer, type);
//            observerMap.put(observer, observerWrapper);
//            liveData.observeForever(observerWrapper);
//        }
//
//        @MainThread
//        private void removeObserverInternal(@NonNull Observer<T> observer) {
//            Observer<T> realObserver;
//            if (observerMap.containsKey(observer)) {
//                realObserver = observerMap.remove(observer);
//            } else {
//                realObserver = observer;
//            }
//            liveData.removeObserver(realObserver);
//        }
//
//        private class LifecycleLiveData<T> extends ExternalLiveData<T> {
//            @Override
//            protected Lifecycle.State observerActiveLevel() {
//                return lifecycleObserverAlwaysActive ? Lifecycle.State.CREATED : Lifecycle.State.STARTED;
//            }
//
//            @Override
//            public void removeObserver(@NonNull Observer<? super T> observer) {
//                super.removeObserver(observer);
//                if (autoClear && !liveData.hasObservers()) {
//                    ZdEvent.get().bus.remove(key);
//                }
//            }
//        }
//
//        private class PostValueTask implements Runnable {
//            private Object newValue;
//
//            public PostValueTask(@NonNull Object newValue) {
//                this.newValue = newValue;
//            }
//
//            @Override
//            public void run() {
//                postInternal((T) newValue);
//            }
//        }
//    }
//
//    private static class ObserverWrapper<T> implements Observer<T> {
//
//        @NonNull
//        private final Observer<T> observer;
//        private Class<T> type;
//        private boolean preventNextEvent = false;
//
//        ObserverWrapper(@NonNull Observer<T> observer, Class<T> type) {
//            this.observer = observer;
//            this.type = type;
//        }
//
//        @Override
//        public void onChanged(@Nullable T t) {
//            if (preventNextEvent) {
//                preventNextEvent = false;
//                return;
//            }
//            try {
//                if (null != type) {
//                    observer.onChanged(new Gson().fromJson(new Gson().toJson(t), type));
//                } else {
//                    observer.onChanged(t);
//                }
//            } catch (ClassCastException e) {
//                e.printStackTrace();
//            }
//        }
//    }
//}

File: platform/core/event/src/main/java/com/android/event/ZdEvent.kt
package com.android.event

import android.os.Handler
import android.os.Looper
import androidx.lifecycle.*
import androidx.lifecycle.Observer
import com.android.event.ZdEvent.ObserverWrapper
import java.util.*
import java.util.Map
import java.util.concurrent.TimeUnit

/**
 * created by jiangshide on 2019-09-29.
 * email:18311271399@163.com
 */
class ZdEvent private constructor(){

  private val bus: MutableMap<String, BusMutableLiveData<Any>>

  init {
    bus = HashMap()
  }

  private object SingletonHolder {
    val DEFAULT_BUS = ZdEvent()
  }

  @Synchronized
  fun <T> with(key: String, type: Class<T>): Observable<T> {
    if (!bus.containsKey(key)) {
      bus[key] = BusMutableLiveData()
    }
    return bus[key] as Observable<T>
  }

  fun with(key: String): Observable<Any> {
    return with(key, Any::class.java)
  }

  interface Observable<T> {

    fun setValue(value: T)

    fun post(value: T)

    fun postValueDelay(value: T, delay: Long, unit: TimeUnit)

    fun observes(owner: LifecycleOwner, observer: Observer<T>)

    fun observeSticky(owner: LifecycleOwner, observer: Observer<T>)

    fun observeForevers(observer: Observer<T>)

    fun observeStickyForever(observer: Observer<T>)

    fun removeObservers(observer: Observer<T>)
  }

  private class BusMutableLiveData<T> : MutableLiveData<T>(), Observable<T> {

    private val observerMap = HashMap<Observer<T>, Observer<T>>()
    private val mainHandler = Handler(Looper.getMainLooper())

    private inner class PostValueTask(private val newValue: T) : Runnable {

      override fun run() {
        value = newValue as T
      }
    }

    override fun post(value: T) {
      mainHandler.post(PostValueTask(value))
    }

    override fun postValueDelay(value: T, delay: Long, unit: TimeUnit) {
      mainHandler.postDelayed(PostValueTask(value), TimeUnit.MILLISECONDS.convert(delay, unit))
    }

    override fun observes(owner: LifecycleOwner, observer: Observer<T>) {
      //保存LifecycleOwner的当前状态
      val lifecycle = owner.lifecycle
      val currentState = lifecycle.currentState
      val observerSize = getLifecycleObserverMapSize(lifecycle)
      val needChangeState = currentState.isAtLeast(Lifecycle.State.STARTED)
      if (needChangeState) {
        //把LifecycleOwner的状态改为INITIALIZED
        setLifecycleState(lifecycle, Lifecycle.State.INITIALIZED)
        //set observerSize to -1，否则super.observe(owner, observer)的时候会无限循环
        setLifecycleObserverMapSize(lifecycle, -1)
      }
      super.observe(owner, observer)
      if (needChangeState) {
        //重置LifecycleOwner的状态
        setLifecycleState(lifecycle, currentState)
        //重置observer size，因为又添加了一个observer，所以数量+1
        setLifecycleObserverMapSize(lifecycle, observerSize + 1)
        //把Observer置为active
        hookObserverActive(observer, true)
      }
      //更改Observer的version
      hookObserverVersion(observer)
    }

    override fun observeSticky(owner: LifecycleOwner, observer: Observer<T>) {
      super.observe(owner, observer)
    }

    override fun observeForevers(observer: Observer<T>) {
      if (!observerMap.containsKey(observer)) {
        observerMap[observer] = ObserverWrapper(observer)
      }
      super.observeForever(observerMap[observer]!!)
    }

    override fun observeStickyForever(observer: Observer<T>) {
      super.observeForever(observer)
    }

    override fun removeObservers(observer: Observer<T>) {
      var realObserver: Observer<T>? = null
      if (observerMap.containsKey(observer)) {
        realObserver = observerMap.remove(observer)
      } else {
        realObserver = observer
      }
      super.removeObserver(realObserver!!)
    }

    private fun setLifecycleObserverMapSize(lifecycle: Lifecycle?, size: Int) {
      if (lifecycle == null) {
        return
      }
      if (lifecycle !is LifecycleRegistry) {
        return
      }
      try {
        val observerMapField = LifecycleRegistry::class.java.getDeclaredField("mObserverMap")
        observerMapField.isAccessible = true
        val mObserverMap = observerMapField.get(lifecycle)
        val superclass = mObserverMap.javaClass.superclass
        val mSizeField = superclass?.getDeclaredField("mSize")
        if (mSizeField != null) {
          mSizeField.isAccessible = true
        }
        mSizeField?.set(mObserverMap, size)
      } catch (e: Exception) {
        e.printStackTrace()
      }

    }

    private fun getLifecycleObserverMapSize(lifecycle: Lifecycle?): Int {
      if (lifecycle == null) {
        return 0
      }
      if (lifecycle !is LifecycleRegistry) {
        return 0
      }
      try {
        val observerMapField = LifecycleRegistry::class.java.getDeclaredField("mObserverMap")
        observerMapField.isAccessible = true
        val mObserverMap = observerMapField.get(lifecycle)
        val superclass = mObserverMap.javaClass.superclass
        val mSizeField = superclass?.getDeclaredField("mSize")
        if (mSizeField != null) {
          mSizeField.isAccessible = true
        }
        return mSizeField?.get(mObserverMap) as Int
      } catch (e: Exception) {
        e.printStackTrace()
        return 0
      }

    }

    private fun setLifecycleState(lifecycle: Lifecycle?, state: Lifecycle.State) {
      if (lifecycle == null) {
        return
      }
      if (lifecycle !is LifecycleRegistry) {
        return
      }
      try {
        val mState = LifecycleRegistry::class.java.getDeclaredField("mState")
        mState.isAccessible = true
        mState.set(lifecycle, state)
      } catch (e: Exception) {
        e.printStackTrace()
      }

    }

    @Throws(Exception::class)
    private fun getObserverWrapper(observer: Observer<T>): Any? {
      val fieldObservers = LiveData::class.java.getDeclaredField("mObservers")
      fieldObservers.isAccessible = true
      val objectObservers = fieldObservers.get(this)
      val classObservers = objectObservers.javaClass
      val methodGet = classObservers.getDeclaredMethod("get", Any::class.java)
      methodGet.isAccessible = true
      val objectWrapperEntry = methodGet.invoke(objectObservers, observer)
      var objectWrapper: Any? = null
      if (objectWrapperEntry is Map.Entry<*, *>) {
        objectWrapper = (objectWrapperEntry as Map.Entry<*, *>).value
      }
      return objectWrapper
    }

    private fun hookObserverVersion(observer: Observer<T>) {
      try {
        //get wrapper's version
        val objectWrapper = getObserverWrapper(observer) ?: return
        val classObserverWrapper = objectWrapper.javaClass.superclass
        val fieldLastVersion = classObserverWrapper?.getDeclaredField("mLastVersion")
        if (fieldLastVersion != null) {
          fieldLastVersion.isAccessible = true
        }
        //get livedata's version
        val fieldVersion = LiveData::class.java.getDeclaredField("mVersion")
        fieldVersion.isAccessible = true
        val objectVersion = fieldVersion.get(this)
        //set wrapper's version
        fieldLastVersion?.set(objectWrapper, objectVersion)
      } catch (e: Exception) {
        e.printStackTrace()
      }

    }

    private fun hookObserverActive(observer: Observer<T>, active: Boolean) {
      try {
        //get wrapper's version
        val objectWrapper = getObserverWrapper(observer) ?: return
        val classObserverWrapper = objectWrapper.javaClass.superclass
        val mActive = classObserverWrapper?.getDeclaredField("mActive")
        if (mActive != null) {
          mActive.isAccessible = true
        }
        mActive?.set(objectWrapper, active)
      } catch (e: Exception) {
        e.printStackTrace()
      }

    }
  }

  private class ObserverWrapper<T>(private val observer: Observer<T>?) : Observer<T> {

    private val isCallOnObserve: Boolean
      get() {
        val stackTrace = Thread.currentThread().stackTrace
        if (stackTrace != null && stackTrace.size > 0) {
          for (element in stackTrace) {
            if ("android.arch.lifecycle.LiveData" == element.className && "observeForever" == element.methodName) {
              return true
            }
          }
        }
        return false
      }

    override fun onChanged(t: T?) {
      if (observer != null) {
        if (isCallOnObserve) {
          return
        }
        observer.onChanged(t)
      }
    }
  }

  companion object {

    fun get(): ZdEvent {
      return SingletonHolder.DEFAULT_BUS
    }
  }
}